### PR TITLE
ai: Handle live-video-to-video requests and start runner

### PR DESCRIPTION
**Pull request by victorges**
**What does this pull request do? Explain your changes. (required)**

This change updates the `/live-video-to-video` handler on the orchestrator to send trickle stream urls to the ai-runner to begin live video AI inference when a live video-to-video AI job is processed

**Specific updates (required)**

- Updates the handler method `StartLiveVideoToVideo()` to send a request to the ai-runner to begin AI inference
- Uses `orch.CheckAICapacity()` to ensure ai-runner is ready for the request
- Only creates trickle channels once job accepted by the ai-runner

**How did you test each of these updates (required)**

Sending requests to gateway with active test stream:
```
curl --location 'http://192.168.10.61:8937/live/video-to-video/test/start' \
--form 'stream="test"' \
--form 'source_id="testout"' \
--form 'source_type="webrtcSession"' \
--form 'query="pipeline=streamdiffusion"'
```

**Does this pull request close any open issues?**

**Checklist:**

- [ ] Read the [contribution guide](./CONTRIBUTING.md)
- [ ] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: server/ai_http.go [hunk 1]
```diff
@@ -140,9 +138,86 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+			pu, err := url.Parse(urlPath)
+			if err != nil {
+				return nil, err
+			}
+			if pu.Hostname() != "" {
+				// url has a hostname already so use it
+				return pu, nil
+			}
+			// no hostname, so append one
+			_, port, err := net.SplitHostPort(r.Host)
```
Comment (j0sh) [inline server/ai_http.go]: this looks sus - the request `r` comes from the gateway. This `appendHostname` dance probably needs to be done on the worker before sending the request down to the runner

File: server/ai_process.go [hunk 2]
```diff
@@ -1043,10 +1044,30 @@ func submitLiveVideoToVideo(ctx context.Context, params aiRequestParams, sess *A
...
+		controlStr := control.String()
+
 		clog.V(common.VERBOSE).Infof(ctx, "pub %s sub %s control %s", pub, sub, control)
 		startTricklePublish(pub, params)
 		startTrickleSubscribe(sub, params)
 		startControlPublish(control, params)
+
+		//Switch subscribe and publish urls
+		req.SubscribeUrl = pub.String() // publish url
+		req.PublishUrl = sub.String() // subscriber url
```
Comment (victorges) [inline server/ai_process.go]: Are these purposefully inverted? Sub = pub and Pub = sub. Maybe it's necessary since the runner args have a different "perspective"

**Review by victorges — COMMENTED** [review]
LGTM!

File: server/ai_http.go (hunk 1 above)
Comment (eliteprox) [reply server/ai_http.go]: Totally agree, I've moved this to `ai_process.go` where it uses `sess.transcoder()`. I will do tests to ensure this works with all different IP types

File: server/ai_process.go (hunk 2 above)
Comment (eliteprox) [reply server/ai_process.go]: Correct, the initial values have `-out` as the subscriber url on the orchestrator so it is necessary to swap these so the ai-runner publishes to `-out`

File: server/ai_process.go [hunk 3]
```diff
@@ -1043,10 +1044,30 @@ func submitLiveVideoToVideo(ctx context.Context, params aiRequestParams, sess *A
...
+		req.ControlUrl = &controlStr
+		
+		// Send another request to the orchestrator to send the video to the live pipeline
```
Comment (j0sh) [inline server/ai_process.go]: I am a little confused why this is needed, can you describe what is happening at a high level maybe in the PR description?

File: server/ai_http.go [hunk 4]
```diff
@@ -86,36 +89,66 @@ func aiHttpHandle[I any](h *lphttp, decoderFunc func(*I, *http.Request) error) h
...
-			})
-		if err != nil {
-			respondWithError(w, err.Error(), http.StatusInternalServerError)
+		var req worker.GenLiveVideoToVideoJSONRequestBody
+		if err := jsonDecoder[worker.GenLiveVideoToVideoJSONRequestBody](&req, r); err != nil {
+			respondWithError(w, err.Error(), http.StatusBadRequest)
 			return
 		}
 
+		// `startStream` is sent on the 2nd request from to the gateway to hand-off the running stream to the ai-runner
```
Comment (j0sh) [inline server/ai_http.go]: I am still not sure why a second request is necessary; the O already has all the parameters  and the G is just sending those back. Why not  submit those params to the worker as part of processing the first request?

File: server/ai_process.go (hunk 3 above)
Comment (victorges) [reply server/ai_process.go]: @eliteprox added on PR desc now!

File: server/ai_http.go (hunk 4 above)
Comment (j0sh) [reply server/ai_http.go]: If you do not want to modify the worker to return params back  to the G (in order to use `handleAIRequest`) then you can copy out the relevant parts from `handleAIRequest` and invoke the required calls directly; AFAICS you're left with only a capacity check and `orch.LiveVideoToVideo`

Most of `handleAIRequest` is boilerplate that can be omitted for now (specialized pipeline handling, payments, metrics)

Comment (eliteprox) [reply server/ai_http.go]: Makes sense, I will modify the flow to exclude `handleAIRequest` and try to do this in a single request. It first seemed there was a context issue with starting the trickle stream before the end of the http request which led to the handoff to fail, but I think `handleAIRequest` was just interfering with the response

Comment (eliteprox) [reply server/ai_http.go]: @j0sh I've updated `StartLiveVideoToVideo` to include the relevant parts from `handleAIRequest`

File: server/ai_http.go [hunk 5]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		workerReq := worker.LiveVideoToVideoParams{
+			ModelId:      req.ModelId,
+			PublishUrl:   req.SubscribeUrl, // SubscribeUrl is the publish url for the worker
+			SubscribeUrl: req.PublishUrl,   // PublishUrl is the subscribe url for the worker
+			ControlUrl:   req.ControlUrl,
+			Params:       req.Params,
+		}
+
+		// append params to the request
+		queryParams := r.FormValue("query")
```
Comment (j0sh) [inline server/ai_http.go]: is this correct? don't the params come in through the request json?

File: server/ai_http.go [hunk 6]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		if params != "" {
+			if err := json.Unmarshal([]byte(params), &paramsMap); err != nil {
+				respondWithError(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+			workerReq.Params = &paramsMap
+		}
+
+		// Send request to the worker
+		_, err = orch.LiveVideoToVideo(ctx, requestID, workerReq)
```
Comment (j0sh) [inline server/ai_http.go]: there is a race condition here if the worker starts subscribing before the channel has been created. the channel needs to be created first. (and then if any of the following requests fail, then the channels need to be closed.)

File: server/ai_http.go [hunk 7]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		}
+
+		orch := h.orchestrator
+		pipeline := "live-video-to-video"
+		cap := core.Capability_LiveVideoToVideo
+		modelID := *req.ModelId
+		clog.V(common.VERBOSE).Infof(ctx, "Received request id=%v cap=%v modelID=%v", requestID, cap, modelID)
+
+		// Create storage for the request (for AI Workers, must run before CheckAICapacity)
+		err = orch.CreateStorageForRequest(requestID)
```
Comment (j0sh) [inline server/ai_http.go]: what is this used for in this pipeline? if its not used, lets remove it.

File: server/ai_http.go [hunk 8]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		}
+
+		// Check if there is capacity for the request
+		if !orch.CheckAICapacity(pipeline, modelID) {
+			respondWithError(w, fmt.Sprintf("Insufficient capacity for pipeline=%v modelID=%v", pipeline, modelID), http.StatusServiceUnavailable)
+			return
+		}
+
+		// Start trickle server for live-video
+		mid := string(core.RandomManifestID())
```
Comment (j0sh) [inline server/ai_http.go]: don't we already have a `requestID` that is created in the exact same way - can that be reused?

File: server/ai_http.go [hunk 9]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		pubUrl := TrickleHTTPPath + mid
```
Comment (j0sh) [inline server/ai_http.go]: nit, but the previous code had a neater style with `var ( ... )`, preference for keeping that as-is.

in general - when doing a self review, look at the diff and try to minimize unnecessary changes

File: server/ai_http.go [hunk 10]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		subUrl := pubUrl + "-out"
+
+		//Append orchestrator service url host to the urls
+		host := h.node.GetServiceURI().String()
+		publishUrl, _ := common.AppendHostname(pubUrl, host)
```
Comment (j0sh) [inline server/ai_http.go]: preference for handling errors instead of ignoring them.

Also, preference for keeping the old behavior of using a relative URL, not appending a host here. Keep that on the G and add on the W instead. I am not sure exactly how much we should rely on orchestrator-local Service URI as opposed to a concrete address that the G/W is already interacting with. for example, W might be local / nearby and we don't want to reach outside the network for that.

File: server/ai_http.go [hunk 11]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		req.PublishUrl = publishUrl.String()
+
+		subscribeUrl, _ := common.AppendHostname(subUrl, host)
+		req.SubscribeUrl = subscribeUrl.String()
+
+		controlUrl, _ := common.AppendHostname(pubUrl+"-control", host)
+		controlUrlStr := controlUrl.String()
+		req.ControlUrl = &controlUrlStr
```
Comment (j0sh) [inline server/ai_http.go]: this is pretty annoying - can we make the `ControlUrl` an optional field that is not [treated as a pointer](https://github.com/livepeer/ai-worker/blob/ca5d650f9ebb60ee4ff8525b4076d19ba93fd169/worker/runner.gen.go#L248-L254)? Eg, an empty / default value would be just fine. Preference for fixing that in the AI worker if possible.

Same goes for the ModelId and Params fields in the LiveVideoToVideoParams

File: server/ai_process.go [hunk 12]
```diff
@@ -1013,41 +1013,25 @@ func submitLiveVideoToVideo(ctx context.Context, params aiRequestParams, sess *A
...
-			if pu.Hostname() != "" {
-				// url has a hostname already so use it
-				return pu, nil
-			}
-			// no hostname, so append one
-			u := sess.Transcoder() + urlPath
-			return url.Parse(u)
-		}
-		pub, err := appendHostname(resp.JSON200.PublishUrl)
+		pub, err := url.Parse(resp.JSON200.PublishUrl)
```
Comment (j0sh) [inline server/ai_process.go]: Should preserve the old behavior here of appending the hostname if necessary, because I am not sure if orchestrators are guaranteed to have a ServiceURI set

File: server/ai_process.go [hunk 13]
```diff
@@ -1013,41 +1013,25 @@ func submitLiveVideoToVideo(ctx context.Context, params aiRequestParams, sess *A
...
 		if err != nil {
-			return nil, fmt.Errorf("pub url - %w", err)
+			return nil, fmt.Errorf("invalid publish URL: %w", err)
 		}
-		sub, err := appendHostname(resp.JSON200.SubscribeUrl)
+		sub, err := url.Parse(resp.JSON200.SubscribeUrl)
```
Comment (j0sh) [inline server/ai_process.go]: as mentioned earlier - preference for not changing this. this will still work even if we have a hostname.

File: server/ai_http.go [hunk 14]
```diff
@@ -96,26 +97,95 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
-			subUrl     = pubUrl + "-out"
-			controlUrl = pubUrl + "-control"
-		)
-		jsonData, err := json.Marshal(
-			&worker.LiveVideoToVideoResponse{
-				PublishUrl:   pubUrl,
-				SubscribeUrl: subUrl,
-				ControlUrl:   controlUrl,
-			})
+		payment, err := getPayment(r.Header.Get(paymentHeader))
```
Comment (j0sh) [inline server/ai_http.go]: we aren't dealing with payments at this point so lets omit this and @leszko can handle that in https://github.com/livepeer/go-livepeer/pull/3261

File: server/ai_process.go (hunk 12 above)
Comment (j0sh) [reply server/ai_process.go]: correcting myself here - looks like it _is_ set but still not sure we should rely on it instead of a relative path from  [this comment](https://github.com/livepeer/go-livepeer/pull/3263#discussion_r1853282905)

File: server/ai_http.go (hunk 14 above)
Comment (eliteprox) [reply server/ai_http.go]: This is being used for logging the eth address of the gateway downstream, is there another way we can get it?

File: server/ai_http.go (hunk 6 above)
Comment (eliteprox) [reply server/ai_http.go]: Reordered to start in the correct order and cleanup publishers channels if job request fails

File: server/ai_process.go (hunk 13 above)
Comment (eliteprox) [reply server/ai_process.go]: Moved the hostname logic so that it uses the original value from `sess.Transcoder()` on the gateway, accurately identifying the remote IP of the orchestrator. It's appended on the orchestrator side of the request processing before being sent to the ai-runner

File: server/ai_process.go (hunk 12 above)
Comment (eliteprox) [reply server/ai_process.go]: Agree, I reverted back to `sess.Transcoder()` from the gateway, see above comment

File: server/ai_http.go (hunk 7 above)
Comment (eliteprox) [reply server/ai_http.go]: This is required for the AI Worker downstream to process the request (I tested without it), It is boilerplate on all the pipelines and might need some refactoring later to make it optional since we are not expecting a response back necessarily.

File: server/ai_http.go (hunk 9 above)
Comment (eliteprox) [reply server/ai_http.go]: I placed this back to how you had it, apologies for that

File: server/ai_http.go (hunk 10 above)
Comment (eliteprox) [reply server/ai_http.go]: Added error handling for url parsing

For the trickle host, I reverted back to `sess.Transcoder()` to get the remote address accurately and then apply it in the ai-runner request. It passes the host through PublishUrl for now

File: server/ai_http.go (hunk 8 above)
Comment (eliteprox) [reply server/ai_http.go]: Re-used it

File: server/ai_http.go (hunk 14 above)
Comment (eliteprox) [reply server/ai_http.go]: Used an empty ETH address for now

Comment (leszko) [reply server/ai_http.go]: yes, I'll handle this in #3261

File: server/ai_http.go [hunk 15]
```diff
@@ -139,8 +155,85 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
 			}
 		}()
 
-		// TODO subscribe to the subscribeUrl for output monitoring
+		// TODO: Use params temporarily to pass the host from gateway
+		host := req.PublishUrl
```
Comment (leszko) [inline server/ai_http.go]: Will the `host` be the same for both runner and gateway?

I think the setup will be like the runner is running inside docker and orchestrator is running directly on the host. Then the publish URL should contain `host.docker.internal`? Or will they both run in the host network?

File: core/ai_worker.go [hunk 16]
```diff
@@ -549,6 +549,39 @@ func (orch *orchestrator) TextToImage(ctx context.Context, requestID string, req
 	return res.Results, nil
 }
 
+func (orch *orchestrator) LiveVideoToVideo(ctx context.Context, requestID string, req worker.GenLiveVideoToVideoJSONRequestBody) (interface{}, error) {
```
Comment (leszko) [inline core/ai_worker.go:552]: I undestand that this is the implementation for the Remote Worker? Have you tested if it works in that setup?

File: server/ai_http.go [hunk 17]
```diff
@@ -139,8 +155,85 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		control, err := common.AppendHostname(controlUrl+"-control", host)
+		if err != nil {
+			respondWithError(w, fmt.Errorf("invalid control URL: %w", err).Error(), http.StatusBadRequest)
+			return
+		}
+
+		// Prepare request to worker
+		workerReq := worker.LiveVideoToVideoParams{
+			ModelId:      req.ModelId,
+			PublishUrl:   sub, // SubscribeUrl is the publish url for the worker
```
Comment (leszko) [inline server/ai_http.go]: `// SubscribeUrl is the publish url for the worker` => I'm not sure I understand this comment. Could you rephrase it?

File: server/ai_http.go (hunk 5 above)
Comment (leszko) [reply server/ai_http.go]: I also don't understand, do we pass all `query` and `params` from Gateway into Orchestrator?

File: server/ai_http.go [hunk 18]
```diff
@@ -139,8 +155,85 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
...
+		}
+
+		// Prepare the response
+		resp := &worker.LiveVideoToVideoResponse{
+			PublishUrl:   workerReq.PublishUrl,
+			SubscribeUrl: workerReq.SubscribeUrl,
+		}
+
+		// TODO: Fix the nullable type in the worker response
+		if workerReq.ControlUrl != nil {
```
Comment (leszko) [inline server/ai_http.go]: What is this? Why ControlUrl is not passed from the worker the same as in the case of `PublishUrl`?

**Review by leszko — COMMENTED** [review]
Added some comments. The main thing is that it didn't work for me, because the address of trickle server is outside Docker and then runner (running inside docker) cannot access the trickle server. Could you please comment on this @victorges @eliteprox ? 

Maybe we need some way of specifying out to access Orchestrator from runner?

File: server/ai_http.go (hunk 18 above)
Comment (eliteprox) [reply server/ai_http.go]: ControlUrl was nullable on the ai-runner, I have solved this in https://github.com/livepeer/ai-worker/pull/286

File: server/ai_http.go (hunk 15 above)
Comment (eliteprox) [reply server/ai_http.go]: I'm having trouble getting this to work correctly. This is why I initially used the ServiceURI of the orchestrator, otherwise we might need to set a startup flag

Comment (j0sh) [reply server/ai_http.go]: Referring back to [this comment](https://github.com/livepeer/go-livepeer/pull/3263#discussion_r1853282905) - this  is why we should stay close to the original behavior of this code and keep the paths generated by the O relative. At the G, append the host name from sess.Transcoder. At the worker, just before sending down the request to the runner (or within the runner), append host.docker.internal as @leszko mentioned (or the O address in the case of remote workers; presumably the worker already has that info in order to connect to the O).

File: core/ai_worker.go (hunk 16 above)
Comment (eliteprox) [reply core/ai_worker.go:552]: I have not yet tested if it works with the remote ai-worker configuration

File: server/ai_http.go (hunk 5 above)
Comment (eliteprox) [reply server/ai_http.go]: I've fixed this now so that it uses the `Params` field from ai_mediaserver.go https://github.com/livepeer/go-livepeer/blob/527bd90fd0e053f9c49741fe124ff3f58c11fca6/server/ai_mediaserver.go#L489-L493

https://github.com/livepeer/go-livepeer/blob/527bd90fd0e053f9c49741fe124ff3f58c11fca6/server/ai_http.go#L163C1-L170C4

File: server/ai_http.go (hunk 11 above)
Comment (eliteprox) [reply server/ai_http.go]: I have updated it to not optional in https://github.com/livepeer/ai-worker/pull/286. I'll re-visit this to address the pointer, gen_openapi.py is requiring a pointer even when default is used

File: server/ai_http.go (hunk 15 above)
Comment (eliteprox) [reply server/ai_http.go]: I have updated this logic from your suggestion and it tested well. On the orchestrator, it will send only the port number to the ai-runner, which uses the client IP from the request to determine the trickle host. https://github.com/livepeer/ai-worker/pull/286/commits/2661dfa31d91caaf0caa900e7b08782d5e9afedf

Passing the trickle port using a temporary param value https://github.com/livepeer/go-livepeer/pull/3263/files#diff-2d409e07b090aec3653ceddfa542b8809eea772f3f24920e8c285d1e22edd15bR157-R161

This tested well across docker containers and with local/remote IPs.

File: server/ai_http.go (hunk 17 above)
Comment (eliteprox) [reply server/ai_http.go]: Updated comment to `Subscribe and Publish urls are swapped for the worker`

**Review by leszko — COMMENTED** [review]
I'm trying to test this PR, but got the following issue, @eliteprox did you maybe happen to encounter it?

```
E1125 15:17:31.015799       1 ai_worker.go:559] clientIP=127.0.0.1 Error processing with local ai worker err="Error response from daemon: invalid mount config for type \"bind\": bind source path does not exist: /models"
2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a
2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a-out
2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a-control
E1125 15:17:31.016013       1 handlers.go:1596] HTTP Response Error statusCode=500 err=Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /models
```

My setup is mostly what we're running in our environment, so dockerized go-livepeer with host network.

**Comment by eliteprox** [comment]
> I'm trying to test this PR, but got the following issue, @eliteprox did you maybe happen to encounter it?
> 
> ```
> E1125 15:17:31.015799       1 ai_worker.go:559] clientIP=127.0.0.1 Error processing with local ai worker err="Error response from daemon: invalid mount config for type \"bind\": bind source path does not exist: /models"
> 2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a
> 2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a-out
> 2024/11/25 15:17:31 INFO Deleted stream streamName=6392c04a-control
> E1125 15:17:31.016013       1 handlers.go:1596] HTTP Response Error statusCode=500 err=Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /models
> ```
> 
> My setup is mostly what we're running in our environment, so dockerized go-livepeer with host network.

If you're running go-livepeer in docker, make sure the volume config is set for `/models`. 
` -v ~/.lpData/:/root/.lpData/`

Also make sure the `-aiModelsDir` is set, otherwise it will default to `/models`     
`-aiModelsDir ~/.lpData/models`

The internal volume mount path should match up with the directory used by `-aiModelsDir`

**Review by victorges — COMMENTED** [review]
LGTM (most of the code is not mine ok :P)

File: server/ai_http.go (hunk 18 above)
Comment (eliteprox) [reply server/ai_http.go]: We've reverted this back to an optional field
